domain/port/state/state_test.go

```diff
@@ -0,0 +1,349 @@
+
+package state
+
+import (
+	"context"
+	"database/sql"
+
+	jc "github.com/juju/testing/checkers"
+	gc "gopkg.in/check.v1"
+
+	"github.com/juju/juju/core/network"
+	"github.com/juju/juju/domain/application"
+	"github.com/juju/juju/domain/application/charm"
+	applicationstate "github.com/juju/juju/domain/application/state"
+	"github.com/juju/juju/internal/changestream/testing"
+	"github.com/juju/juju/internal/logger"
+)
+
+type stateSuite struct {
+	testing.ModelSuite
+
+	unitUUID string
+}
+
+var _ = gc.Suite(&stateSuite{})
+
+func ptr[T any](v T) *T {
+	return &v
+}
+
+func (s *stateSuite) SetUpTest(c *gc.C) {
+	s.ModelSuite.SetUpTest(c)
+	applicationSt := applicationstate.NewState(s.TxnRunnerFactory(), logger.GetLogger("juju.test.application"))
+	_, err := applicationSt.CreateApplication(context.Background(), "app", application.AddApplicationArg{
+		Charm: charm.Charm{
+			Metadata: charm.Metadata{
+				Name: "app",
+			},
+		},
+	}, application.AddUnitArg{UnitName: ptr("app/0")})
+	c.Assert(err, jc.ErrorIsNil)
+
+	err = s.TxnRunner().StdTxn(context.Background(), func(ctx context.Context, tx *sql.Tx) error {
+		err := tx.QueryRowContext(ctx, "SELECT uuid FROM unit").Scan(&s.unitUUID)
+		c.Assert(err, jc.ErrorIsNil)
+		return nil
+	})
+	c.Assert(err, jc.ErrorIsNil)
+}
+
+func (s *stateSuite) initialiseOpenPort(c *gc.C, st *State) {
+	ctx := context.Background()
+	err := st.UpdateUnitPorts(ctx, s.unitUUID, network.GroupedPortRanges{
+		"endpoint": {
+			{Protocol: "tcp", FromPort: 80, ToPort: 80},
+			{Protocol: "udp", FromPort: 1000, ToPort: 1500},
+		},
+		"misc": {
+			{Protocol: "tcp", FromPort: 8080, ToPort: 8080},
+		},
+	}, network.GroupedPortRanges{})
+	c.Assert(err, jc.ErrorIsNil)
+}
+
+func (s *stateSuite) TestGetOpenedPortsBlankDB(c *gc.C) {
+	st := NewState(s.TxnRunnerFactory())
+	ctx := context.Background()
+
+	groupedPortRanges, err := st.GetOpenedPorts(ctx, s.unitUUID)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Check(groupedPortRanges, gc.HasLen, 0)
+
+	groupedPortRanges, err = st.GetOpenedPorts(ctx, "non-existent")
+	c.Assert(err, jc.ErrorIsNil)
+	c.Check(groupedPortRanges, gc.HasLen, 0)
+}
+
+func (s *stateSuite) TestGetOpenedPorts(c *gc.C) {
+	st := NewState(s.TxnRunnerFactory())
+	ctx := context.Background()
+	s.initialiseOpenPort(c, st)
+
+	groupedPortRanges, err := st.GetOpenedPorts(ctx, s.unitUUID)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Check(groupedPortRanges, gc.HasLen, 2)
+
+	c.Check(groupedPortRanges["endpoint"], gc.HasLen, 2)
+	c.Check(groupedPortRanges["endpoint"][0], jc.DeepEquals, network.PortRange{Protocol: "tcp", FromPort: 80, ToPort: 80})
+	c.Check(groupedPortRanges["endpoint"][1], jc.DeepEquals, network.PortRange{Protocol: "udp", FromPort: 1000, ToPort: 1500})
+
+	c.Check(groupedPortRanges["misc"], gc.HasLen, 1)
+	c.Check(groupedPortRanges["misc"][0], jc.DeepEquals, network.PortRange{Protocol: "tcp", FromPort: 8080, ToPort: 8080})
+}
+
+func (s *stateSuite) TestOpenPort(c *gc.C) {
```

Comment:
Might be good to have test names start with UpdateUnitPorts since that this the func they are testing.

Though fair enough if you'd rather not, it'll be quite verbose

Reply:
A bit of verbosity is fine. Changed